### PR TITLE
Feature pruning

### PR DESCRIPTION
**Pull request by cowchipkid**
The feature pruning implementation, prunes low value features for SVM, LTU subclasses and sparse nets.

### CODE REVIEW

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java [hunk 1]
```diff
@@ -880,53 +880,57 @@ public void run() {
...
+                    }
+                    
+                    if (lce.parser != null && testParser != null) {
+                        System.out.println("Testing " + getName());
+                        new Accuracy(true).test(learner, learner.getLabeler(), testParser);
+                    }
+                    
+                    // save the final model.
+                    System.out.println("Writing " + getName());
                     learner.save(); // Doesn't write .lex if lexicon is empty.
```
Comment (danyaljj) [inline lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java]: @cowchipkid can you clarify the change here? Confused about what's happening here ....

Comment (cowchipkid) [reply lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java]: I moved the Accuracy reporting outside the block to ensure doneTraining() gets called before accuracy reporting. doneTraining will apply the feature optimization, we want the score after that. If accuracy is reported before doneTraining, we will get the score of the un-optimized models, which would be for not.

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/featurepruning/SparseNetworkOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                if (!whitelist.contains(entry.getKey())) {
+                    int i = 0;
+                    for (; i < numberclasses; ++i) {
+                        LinearThresholdUnit ltu = (LinearThresholdUnit) networkLearner.getNetwork().get(i);
+                        if (ltu == null) {
+                            System.out.println("THERE WAS NO LTU AT " + i);
+                            continue;
+                        }
+                        double wt = ltu.getWeightVector().getRawWeights().get(fi);
+
```
Comment (cowchipkid) [inline lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/featurepruning/SparseNetworkOptimizer.java]: Ya, this is wrong. We need to call hasWeight here I think, this won't work for SparseAveragedPerceptron where we need to sum the past average with the actual weight.

Comment (cowchipkid) [reply lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/featurepruning/SparseNetworkOptimizer.java]: Also, it doesn't even address absolute zero.

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java (hunk 1 above)
Comment (danyaljj) [reply lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java]: Ah I see. Makes sense.

**Comment by cowchipkid** [comment]
The documentation in the feature pruning stuff is in the package-info.java in the package with all the optimization stuff. It was was already in place, I enhanced it a bit, and added some links into the life cycle methods. I also fixed one bug that likely had no impact. It would not disable pruning when you would set the threshold to zero.

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/Lexicon.java [hunk 3]
```diff
@@ -742,10 +775,9 @@ public int compare(int i1, int i2) {
         ByteString previousBSIdentifier = null;
         out.writeInt(indexes.length);
         out.writeInt(pruneCutoff);
-
         for (int i = 0; i < indexes.length; ++i) {
             Feature f = inverse.get(indexes[i]);
-            previousClassName =
+             previousClassName =
```
Comment (danyaljj) [inline lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/Lexicon.java]: drop the extra space?

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SupportVectorMachine.java [hunk 4]
```diff
@@ -72,6 +77,11 @@
     public static final double defaultEpsilon = 0.1;
     /** Default for {@link #bias}. */
     public static final double defaultBias = 1.0;
+    /** any weight less than this is considered irrelevant. This is for prunning. */
+    public static final double defaultFeaturePruningThreshold = 0.000001;
```
Comment (danyaljj) [inline lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SupportVectorMachine.java]: Is this always used in an absolute sense? I.e. it always have to be positive. 
In other words, if its zero, no pruning will be done.

**Comment by danyaljj** [comment]
A minor comment; Looks good to me! 
Unless @mssammon plans to go through this, feel free to merge it.

**Comment by mssammon** [comment]
am still in the thick of post-move activity; feel free to merge.

File: lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SupportVectorMachine.java (hunk 4 above)
Comment (cowchipkid) [reply lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SupportVectorMachine.java]: @danyaljj setting the threshold to zero is how you disable pruning, effectively.

**Pull request merged**